File: Difficulty: Medium/Spirally traversing a matrix/spirally-traversing-a-matrix.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
    vector<int> spirallyTraverse(vector<vector<int> > &matrix) {
        // code here
        int row=matrix.size();
        int col=matrix[0].size();

        int top = 0 , bottom = row - 1  ,left = 0  , right = col - 1 ;
        vector<int> result;

        while (left <= right && top <= bottom) {
            //print top row
            for(int i = left;i<=right;i++){
               result.push_back(matrix[top][i]);
            }
            top++;
            
            //print right column
            for(int i = top;i<=bottom;i++){
                result.push_back(matrix[i][right]);
            }
            right--;
            
            
            if(top<=bottom){
                for(int i = right;i>=left;i--){
                    result.push_back(matrix[bottom][i]);
                }
                bottom--;
                    
            }
            
            if(left<=right){
                for(int i = bottom;i>=top;i--){
                    result.push_back(matrix[i][left]);
                }
                left++;
                    
            }
        }
        return result;
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;

    while (t--) {
        int r, c;
        cin >> r >> c;
        vector<vector<int>> matrix(r, vector<int>(c, 0));

        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                cin >> matrix[i][j];
            }
        }

        Solution ob;
        vector<int> result = ob.spirallyTraverse(matrix);
        for (int i = 0; i < result.size(); ++i)
            cout << result[i] << " ";
        cout << endl;
    }
    return 0;
}
// } Driver Code Ends

File: Difficulty: Medium/Spirally traversing a matrix/spirally-traversing-a-matrix.py
class Solution:
    # Function to return a list of integers denoting spiral traversal of matrix.
    def spirallyTraverse(self, mat):
        # code here
        n = len(mat)
        m = len(mat[0])
        top, left = 0, 0
        bottom, right = n - 1, m - 1
        result = []

        while top <= bottom and left <= right:
            for i in range(left, right + 1):
                result.append(mat[top][i])
            top += 1
            for i in range(top, bottom + 1):
                result.append(mat[i][right])
            right -= 1

            if top <= bottom:
                for i in range(right, left - 1, -1):
                    result.append(mat[bottom][i])
                bottom -= 1

            if left <= right:
                for i in range(bottom, top - 1, -1):
                    result.append(mat[i][left])
                left += 1

        return result


#{ 
 # Driver Code Starts
if __name__ == "__main__":
    import sys
    input = sys.stdin.read
    data = input().split()

    t = int(data[0])
    index = 1
    for _ in range(t):
        r = int(data[index])
        c = int(data[index + 1])
        index += 2
        matrix = []
        for i in range(r):
            row = list(map(int, data[index:index + c]))
            matrix.append(row)
            index += c

        solution = Solution()
        result = solution.spirallyTraverse(matrix)
        print(" ".join(map(str, result)))
        print("~")

# } Driver Code Ends